drivers/flash/flash_gd32.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2022 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_flash_controller
+
+#include <errno.h>
+#include <kernel.h>
+#include <devicetree.h>
+#include <drivers/flash.h>
+#include <soc.h>
+
+#include "flash_gd32.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(flash_gd32, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_gd32_data {
+	struct k_sem mutex;
+};
+
+static struct flash_gd32_data flash_data;
+
+static const struct flash_parameters flash_gd32_parameters = {
+	.write_block_size = SOC_NV_FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+__weak int flash_gd32_programming(off_t offset, const void *data, size_t len)
```

Comment:
@nandojve Please add a DMN for this PR.
I use a weak implement to avoid compile time error. But I think there should have a better solution.

Reply:
You can tell people to not merge by adding a DNM at PR.

Reply:
You can drop all weak functions.

Reply:
use GD32_HAS_FLASH_DRIVER, done.